service/matching/taskqueuemonitor.go

```diff
@@ -0,0 +1,116 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package matching
+
+import (
+	"context"
+	"math"
+	"time"
+)
+
+type (
+	tqmonStop    struct{}
+	tqmonGC      struct{}
+	tqmonUsurped struct{ id *taskQueueID }
+
+	// tqmonAccess is the interface by which a taskQueueMonitor observes and
+	// updates a set of taskQueueManager instances.
+	tqmonAccess interface {
+		getTaskQueues(int) []taskQueueManager
+		unloadTaskQueue(*taskQueueID)
+	}
+
+	// taskQueueMonitor performs garbage collection and unloads instances of
+	// taskQueueManager accessed via a tqmonAccess.
+	taskQueueMonitor struct {
+		sigs chan interface{}
+	}
+)
+
+// newTaskQueueMonitor constructs a new instance of taskQueueMonitor.
+func newTaskQueueMonitor() *taskQueueMonitor {
+	return &taskQueueMonitor{sigs: make(chan interface{})}
+}
+
+// Run runs the taskQueueMonitor event loop, performing GC of a set of
+// taskQueueManager instances and unloading the same when signaled or when GC
+// deems it appropriate. The implementaion of tqmonAccess MUST NOT re-enter this
+// taskQueueMonitor instance.
+func (m *taskQueueMonitor) Run(a tqmonAccess, c *Config) {
+	defer close(m.sigs)
+	t := time.NewTimer(c.IdleTaskqueueCheckInterval())
+	defer t.Stop()
+
+	for {
+		select {
+		case i := <-m.sigs:
+			switch sig := i.(type) {
+			case tqmonUsurped:
+				a.unloadTaskQueue(sig.id)
+			case tqmonGC:
+				m.runGC(a)
+				t.Reset(c.IdleTaskqueueCheckInterval())
+			case tqmonStop:
+				return
+			}
+		case <-t.C:
+			m.runGC(a)
+			t.Reset(c.IdleTaskqueueCheckInterval())
+		}
+	}
+}
+
+// Stop synchronously stops a running taskQueueMonitor. Once stopped a
+// taskQueueMonitor must not be restarted. The only time this function returns
+// an error is if the provided context ends while the shutdown is in process, in
+// which case the ctx.Err() is returned and the internal state of the
+// taskQueueMonitor instance is undefined.
+func (m *taskQueueMonitor) Stop(ctx context.Context) error {
+	select {
+	case m.sigs <- tqmonStop{}:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+	select {
+	case <-m.sigs:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+	return nil
```

Comment:
what is the contract / guarantee?

```
// which case the ctx.Err() is returned and the internal state of the
// taskQueueMonitor instance is undefined.
```

out of curiosity, can the state of task queue monitor be guaranteed after shutdown?

Reply:
Not if I want Stop() to be interruptible (and I think it should be). GC takes microseconds so absent a deadlock I don't really expect a problem. You could say making the channel bigger would work but however big it is, the tqmonStop signal could always be the signal that doesn't fit in the buffer.